rasa/utils/train_utils.py

```diff
@@ -181,78 +268,132 @@ def balance_session_data(
             if min(num_data_cycles) > 0:
                 break
 
-    return SessionData(
-        X=np.concatenate(new_X),
-        Y=np.concatenate(new_Y),
-        label_ids=np.concatenate(new_label_ids),
-    )
+    new_X = {k: np.concatenate(v) for k, v in new_X.items()}
+    new_Y = {k: np.concatenate(v) for k, v in new_Y.items()}
+    new_labels = {k: np.concatenate(v) for k, v in new_labels.items()}
+
+    return SessionData(X=new_X, Y=new_Y, labels=new_labels)
+
+
+def get_number_of_examples(session_data: SessionData):
+    example_lengths = [v.shape[0] for v in session_data.X.values()]
+
+    # check if number of examples is the same for all X
+    if not all(length == example_lengths[0] for length in example_lengths):
+        raise ValueError(
+            f"Number of examples differs for X ({session_data.X.keys()}). There should "
+            f"be the same."
+        )
+
+    return example_lengths[0]
 
 
 def gen_batch(
     session_data: "SessionData",
     batch_size: int,
+    label_key: Text,
     batch_strategy: Text = "sequence",
     shuffle: bool = False,
-) -> Generator[Tuple["np.ndarray", "np.ndarray"], None, None]:
+) -> Generator[Tuple, None, None]:
     """Generate batches."""
 
     if shuffle:
         session_data = shuffle_session_data(session_data)
 
     if batch_strategy == "balanced":
-        session_data = balance_session_data(session_data, batch_size, shuffle)
+        session_data = balance_session_data(
+            session_data, batch_size, shuffle, label_key
+        )
 
-    num_batches = session_data.X.shape[0] // batch_size + int(
-        session_data.X.shape[0] % batch_size > 0
-    )
+    num_examples = get_number_of_examples(session_data)
+    num_batches = num_examples // batch_size + int(num_examples % batch_size > 0)
 
     for batch_num in range(num_batches):
-        batch_x = session_data.X[batch_num * batch_size : (batch_num + 1) * batch_size]
-        batch_y = session_data.Y[batch_num * batch_size : (batch_num + 1) * batch_size]
+        start = batch_num * batch_size
+        end = start + batch_size
 
-        yield batch_x, batch_y
+        batch_data = [sparse_to_dense(v[start:end]) for v in session_data.X.values()]
+        batch_data = batch_data + [
+            sparse_to_dense(v[start:end]) for v in session_data.Y.values()
+        ]
+        batch_data = batch_data + [
+            sparse_to_dense(v[start:end]) for v in session_data.labels.values()
+        ]
+
+        yield tuple(batch_data)
+
+
+def sparse_to_dense(
+    examples: Union[np.ndarray, List[scipy.sparse.csr_matrix]]
```

Comment:
let's add a comment that in case of bow this will be either 2D dense array or list of sparse matrices 1xN (because sparse vector doesn't exist) and in case of sequence it'll be 3D dense array or a list of sparse matrices seq_lenxN

This seems to me a bit unintuitive, but I agree, if we do it in classifier, it makes it more general.